### ISSUE TITLE
Custom Keras model and layers demonstrate different behavior (nonzero vs zero grads) on CPU vs GPU

### ISSUE THREAD

**Issue by jkr26** [tensorflow/tensorflow#39610]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): Tested against 2.1.0 and 2.2.0 GPU vs CPU, as well as TF-nightly 2.3.0.dev20200516 on GPU
- Python version: 3.6.9
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA 10.1, cuDNN 7.6.5
- GPU model and memory: GeForce GTX 1070 Ti

**Current, expected behavior, best repro I can get**
Hi TF core,

Writing some custom layers and models for an at-home research project, primarily using Keras, but also some lower-level TF constructs to programatically construct a few matrices. I am in an interesting position in this project currently, where I can flip between my sanity-check tests passing and failing by *changing the TF version backing my Python interpreter*, specifically by using a TF-GPU version versus a pure TF-CPU version. To be clear, all the gradients in my custom model are 0 (0, not `None`) when using TF-GPU and nonzero when using TF-CPU.

I can inline the tests whose pass and failure can be flipped in this way, but pulling an entire standalone example might be a nontrivial task due to the layering The failing test is as follows:

```python
  def test_grads_non_zero(self):
    model = classification_models.TwoLayerLinalgConv(1, 784)
    with tf.GradientTape() as g:
      result = model(MNIST_DATA)
      loss = tf.keras.losses.sparse_categorical_crossentropy(MNIST_CLASSES, result, from_logits=True)
    grads = g.gradient(loss, model.trainable_weights)
    for idx, grad_elem in enumerate(grads):
      self.assertFalse(grad_elem is None)
      self.assertNotEqual(self.evaluate(tf.reduce_sum(tf.math.abs(grad_elem))), 0.)
```

The model which is used here is a simple composition (via subclassing of `tf.keras.Model`), as follows:

```python
class TwoLayerLinalgConv(tf.keras.Model):
  """Simple 2-dimensional convolution-like model via custom layer."""

  def __init__(
      self,
      n_filters,
      data_size,
      # ...other things
      ):
    super(TwoLayerLinalgConv, self).__init__()
    # Some initializations
    # etc
    self.conv1 = layers.LaplacianLayer(n_filters=self._n_filters,
                                       n_input_channels=self._n_input_channels,
                                       n_params_to_fit=self._kernel_size,
                                       n_dimensions=self._n_dimensions)
        # Must have 1 filter here for the dense layer in the end.
    self.conv2 = layers.LaplacianLayer(n_filters=1,
                                       n_input_channels=n_filters,
                                       n_params_to_fit=self._kernel_size,
                                       n_dimensions=self._n_dimensions)
    self.dense = tf.keras.layers.Dense(output_dim, activation=None)

  def call(self, inputs):
    x = tf.nn.relu(self.conv1(tf.reshape(inputs, [-1, self._n_input_channels, self._data_size])))
    x = tf.nn.relu(self.conv2(x))
    return self.dense(tf.reshape(x, [-1, self._data_size]))
```

It is probably worth noting that the layers used in this model, when tested as follows, actually *pass* with CPU and GPU-backed environments:

```python
  def test_operator_trains(self):
    l = layers.LaplacianLayer(1, 1, 1)
    l.build([10, 1, MNIST_SIZE])
    with tf.GradientTape() as g:
      input_with_channel = tf.reshape(MNIST_DATA, [-1, 1, MNIST_SIZE])
      result_of_layer = l(input_with_channel)
      pred = tf.keras.layers.Dense(10)(result_of_layer)
      result = tf.reduce_mean(tf.keras.losses.sparse_categorical_crossentropy(
          MNIST_LABELS, tf.reshape(pred, shape=[10, 10]), from_logits=True))

    grad = g.gradient(result, l.trainable_weights)
    for grad_elem in grad:
      self.assertFalse(grad_elem is None)
      self.assertNotEqual(self.evaluate(tf.reduce_sum(tf.math.abs(grad_elem))), 0.)
```

I am wondering: is this kind of divergence a known issue? Is there a simple way to mitigate this problem? Is there any more information that would be helpful, or pointers on debugging? I am pretty much at a loss on where to go from here...

Thanks!
Keith

**Comment by Saduf2019** [comment]
@jkr26 
I ran the code shared above and do not see any output to compare, please find the gist for [nightly](https://colab.sandbox.google.com/gist/Saduf2019/a28645812ad436dc89fc5940f9656baf/untitled183.ipynb) and [tf 2.1](https://colab.sandbox.google.com/gist/Saduf2019/4a55ecb963b9afe6565421b3a10c3fa7/untitled182.ipynb).
Can you please share a colab gist for the reported issue, error logs for us to analyse.

**Comment by jkr26** [comment]
@Saduf2019 

Yes, the code above is not entirely self-contained--it would be rather nontrivial to produce such a complete example, but doable. Is this what you would prefer?

**Comment by goldiegadde** [comment]
@jkr26 it would be very helpful if you can provide a reproducible colab gist of the issue. Thanks!
